kolibri/core/notifications/test/test_api.py

```diff
@@ -168,22 +200,232 @@ def test_create_notification(self):
+            self.user1.id,
+            self.classroom.id,
+            assignment_collections=[self.classroom.id],
+            contentnode_id=self.node_1.id,
+            lesson_id=self.lesson.id,
+            timestamp=self.summarylog1.completion_timestamp,
+        )
+
+    @patch("kolibri.core.notifications.api.create_notification")
+    def test_parse_summarylog_not_save_completed_lesson_notification(
+        self, create_notification
+    ):
+        self.summarylog1.progress = 1.0
+        parse_summarylog(self.summarylog1)
+
+        # dont save completed lesson if there are more resources to complete
+        notifications = create_notification.call_args_list
+        assert not any(
+            [
+                (
+                    call[1] == NotificationEventType.Completed
+                    and call[0] == NotificationObjectType.Lesson
+                )
+                for call in notifications
+            ]
+        )
+
+    @patch("kolibri.core.notifications.api.create_notification")
+    @patch("kolibri.core.notifications.api.save_notifications")
+    def test_parse_summarylog_save_completed_lesson_notification(
+        self, save_notifications, create_notification
+    ):
+        self.summarylog1.progress = 1.0
+        self.summarylog1.completion_timestamp = local_now() + timedelta(seconds=2)
+        self.summarylog1.end_timestamp = local_now() + timedelta(seconds=5)
+        self.summarylog1.save()
+        parse_summarylog(self.summarylog1)
+
+        self.summarylog2.progress = 1.0
+        self.summarylog2.completion_timestamp = local_now() + timedelta(seconds=7)
+        self.summarylog2.end_timestamp = local_now() + timedelta(seconds=10)
+        self.summarylog2.save()
+        parse_summarylog(self.summarylog2)
+
+        assert save_notifications.called
+        create_notification.assert_any_call(
+            NotificationObjectType.Lesson,
+            NotificationEventType.Completed,
+            self.user1.id,
+            self.classroom.id,
+            assignment_collections=[self.classroom.id],
+            lesson_id=self.lesson.id,
+            # End timestamp should be the latest completion time
+            timestamp=self.summarylog2.completion_timestamp,
+        )
+
+    @patch("kolibri.core.notifications.api.create_notification")
+    @patch("kolibri.core.notifications.api.save_notifications")
+    def test_parse_summarylog_save_completed_lesson_notification_with_retry(
+        self, save_notifications, create_notification
+    ):
+        self.summarylog1.progress = 1.0
+        self.summarylog1.completion_timestamp = local_now() + timedelta(seconds=2)
+        self.summarylog1.end_timestamp = local_now() + timedelta(seconds=5)
+        self.summarylog1.save()
+        parse_summarylog(self.summarylog1)
+
+        # Student retried the exercise, and even with progress = 0, this should
+        # be considered as completed
+        self.summarylog1.progress = 0
```

Comment:
Hrm - I feel like this should only be the case when we can also independently confirm that there's a masterylog that does have the completion registered?

Would be good to duplicate this test as is, and assert that it doesn't set the lesson as complete - and then in the other copy add a complete and an incomplete masterylog, and assert that it does set the lesson as complete.

Reply:
Oh, here I'm relying more on the assumption that if the completion_timestamp has a value (we set it [here](https://github.com/AlexVelezLl/kolibri/blob/39c612040566b612fd0bef4ca3d47d90e17d5d2e/kolibri/core/notifications/test/test_api.py#L284)), it means that at some point it was completed, and if it has progress 0 and completion_timestamp != null, then it's because there was a retry.

Reply:
I updated the test to include the masterylog to make it more attached to reality, but I did not add the assert test that a lesson notification was not generated with an uncompleted resource, because that would be being tested https://github.com/AlexVelezLl/kolibri/blob/39c612040566b612fd0bef4ca3d47d90e17d5d2e/ kolibri/core/notifications/test/test_api.py#L216.

Reply:
OK, we can keep this as is for now, I think I need to mull further on what the implications for this might be - but it's not a regression by any means!